### PR TITLE
Streamlit app development

### PR DESCRIPTION
**Pull request by mkutarna**
Update master branch with streamlit app work.
- App accepts more input files (pdf, txt, epub)
- Better text cleaning to reduce tts engine errors
- Improved output filing handling

### CODE REVIEW

File: tests/data/test.htm [hunk 1]
```diff
@@ -0,0 +1,118 @@
+<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
```
Comment (rkingery) [inline tests/data/test.htm:1]: Make sure all `__pycache__` folders are in the `.gitignore`. You don't wanna see these on github. If they are in the gitignore (formatted correctly) but still showing up on github, then explicitly delete those files/folders from the remote repo.

File: requirements.txt [hunk 2]
```diff
@@ -0,0 +1,7 @@
+silero
```
Comment (rkingery) [inline requirements.txt:1]: Make sure this file contains the minimal working set of libraries you use in your repo. If anybody tries to do a `pip install -r requirements.txt`, it should download *every* library you use, in the right versions, and they should work in every file.

File: instructions.md [hunk 3]
```diff
@@ -0,0 +1,13 @@
+This tool generates custom-voiced audiobook files from an imported ebook file. Please upload an ebook to begin the conversion process. Output files will be downloaded as a .zip archive.
```
Comment (rkingery) [inline instructions.md]: Make this a section in your README, not a separate file. To see what sections your README should contain, see my python packages notes on Slack.

File: src/config.py [hunk 4]
```diff
@@ -0,0 +1,12 @@
+DEVICE = 'cpu'
+LANGUAGE = 'en'
+MAX_CHAR_LEN = 140
+MODEL_ID = 'v3_en'
+SAMPLE_RATE = 24000
+# SPEAKER_LIST = ['en_0', 'en_29', 'en_41', 'en_110']
```
Comment (rkingery) [inline src/config.py]: Don't include commented code in source code. Either keep or delete.

Comment (rkingery) [reply src/config.py]: Your config should probably contain any file paths you need to use anywhere in your code as well. Never hardcode file paths since they can easily change in a reorg.

File: app.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        text = preprocess(file)
+    elif file_ext == 'application/pdf':
+        text = read_pdf(uploaded_file)
+    elif file_ext == 'text/html' or file_ext == 'text/htm':
+        ebook_upload = uploaded_file.read()
+        text = read_html(ebook_upload)
+    else:
+        st.warning('Invalid file type', icon="⚠️")
+    st.success('Reading file complete!')
+    # st.text(text)
```
Comment (rkingery) [inline app.py]: Delete or keep commented out source code. Never leave in commented out form.

File: src/predict.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+from stqdm import stqdm
+
+from . import config as cf
+from src.output import write_audio
+
+def load_model():
+    from silero import silero_tts
+
+    model, _ = silero_tts(language=cf.LANGUAGE,
+                                speaker=cf.MODEL_ID)
```
Comment (rkingery) [inline src/predict.py]: You're free to ignore linewidth limits like this. Feel free to take up to screen width per line (say 150 characters per line). The old 70 character limit is kind of old-fashioned now.

File: src/read.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
+import re
```
Comment (rkingery) [inline src/read.py]: Using `read.py` is kind of vague. Maybe something like `file_readers.py` would be better. Filenames should describe as succinctly and as informatively as possible what the objects inside that file do.

File: src/read.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    corpus = []
+    for item in stqdm(list(reader.pages), desc="Pages in pdf:"):
+        text_list = preprocess(item.extract_text())
+        corpus.append(text_list)
+
+    return corpus
+
+def read_html(file):
+    text_list = preprocess(file)
+
```
Comment (rkingery) [inline src/read.py]: Don't leave unnecessary spaces like this. For short functions (less than say 5 lines), no point in using any spaces. There's no rule about spacing before a return.

File: tests/test_read.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    file = open(txt_path)
+    corpus = rd.preprocess(file)
+    file.close()
+
+    assert np.shape(corpus) == (26,)
+    assert corpus[0] == 'Testing Text File'
+    assert corpus[2] == 'Link to generator repo!'
+    assert corpus[20] == 'Here are some Chinese characters: .'
+    assert corpus[22] == 'The vowels: are , , , , , , .'
+    assert corpus[24] == 'We can also test from mathematical symbols: , , , , , X, %,  ,a, , , +, = ,-.'
```
Comment (rkingery) [inline tests/test_read.py]: Looks like you have a syntax error in your test file. Review that to make sure it's grammatically correct. E.g., this should probably be "for" not "from".

File: tests/test_read.py [hunk 10]
```diff
@@ -0,0 +1,39 @@
+import pytest
+import pathlib
+import numpy as np
+
+import src.read as rd
+
+def test_preprocess():
+    """
```
Comment (rkingery) [inline tests/test_read.py]: Expect these docstrings to change soon as we use a standard, more informative format for them.

File: tests/test_read.py [hunk 11]
```diff
@@ -0,0 +1,39 @@
+import pytest
```
Comment (rkingery) [inline tests/test_read.py]: Looks like you're still missing tests for other file formats (pdf, txt, html). You'll want to get those in here too.

File: tests/test_predict.py [hunk 12]
```diff
@@ -0,0 +1,14 @@
+import pytest
```
Comment (rkingery) [inline tests/test_predict.py:1]: You'll want to test predictions as well, by passing through a standard piece of text and making sure the output tensor looks like it's supposed to. Just some simple ones will do, and remember edge cases (like empty text, or text from multiple chapters, etc).

**Review by rkingery — COMMENTED** [review]
See specific comments. Also, create a card to properly fill in the documentation (README, docstrings, file headers, etc). See my python package notes for advice on how to do these things.

File: src/config.py (hunk 4 above)
Comment (mkutarna) [reply src/config.py]: Commented lines removed.

File: app.py (hunk 5 above)
Comment (mkutarna) [reply app.py]: Commented lines in app.py removed.

File: tests/data/test.htm (hunk 1 above)
Comment (mkutarna) [reply tests/data/test.htm:1]: Checked: `__pycache__` folders are in the `.gitignore` and do not appear on github on this branch.

File: instructions.md (hunk 3 above)
Comment (mkutarna) [reply instructions.md]: README update according to the python packages file. Instructions.md moved to resources folder.

File: src/predict.py (hunk 6 above)
Comment (mkutarna) [reply src/predict.py]: Checked: several lines changed to follow this convention - will follow it going forward.

File: src/read.py (hunk 7 above)
Comment (mkutarna) [reply src/read.py]: Renamed to `file_readers.py` (and refactored other files to follow).

File: src/read.py (hunk 8 above)
Comment (mkutarna) [reply src/read.py]: Checked: removed some un-needed spaces in a few places.

File: tests/test_read.py (hunk 9 above)
Comment (mkutarna) [reply tests/test_read.py]: Checked: I read through the test files and checked for grammar, spelling, etc.

File: tests/test_read.py (hunk 10 above)
Comment (mkutarna) [reply tests/test_read.py]: The docstrings have been implemented, in the recommended style.

File: tests/test_read.py (hunk 11 above)
Comment (mkutarna) [reply tests/test_read.py]: Tests for read_pdf & read_txt implemented. read_html and its tests will be implemented in the next cycle.

File: requirements.txt (hunk 2 above)
Comment (mkutarna) [reply requirements.txt:1]: I've tested the requirements file by creating a new conda environment, running `pip install -r requirements.txt`, and running the app: this works with the current file.

File: tests/test_predict.py (hunk 12 above)
Comment (mkutarna) [reply tests/test_predict.py:1]: I was able to write a test for the predict function by manually setting the seed in the test case.

File: resources/instructions.md [hunk 13]
```diff
@@ -0,0 +1,13 @@
+This tool generates custom-voiced audiobook files from an imported ebook file. Please upload an ebook to begin the conversion process. Output files will be downloaded as a .zip archive.
```
Comment (rkingery) [inline resources/instructions.md:1]: This should really be part of your README. Your project isn't complex enough to need multiple files of documentation. Just put this info at the top of the README after the initial intro.

File: src/file_readers.py [hunk 14]
```diff
@@ -0,0 +1,123 @@
+"""
+Notes
+-----
+This module contains the functions for audiobook_gen that read in the
+file formats that require for parsing than plain text (pdf, html, epub),
+as well as the preprocessing function for all input files.
+"""
+
+__all__ = ['preprocess', 'read_pdf', 'read_epub']
```
Comment (rkingery) [inline src/file_readers.py]: You don't need a `__all__` (pronounced "dunder all") here or in any of your modules since you're already using all these functions. The `__all__` is only for when you want to hide functionality (i.e. functions, classes, variables) in the file from a user importing them. In your case, you're the only one importing them and using them (in your app) so it doesn't matter. This is more relevant for python packages, which yours is not.

File: src/file_readers.py [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+from textwrap import TextWrapper
+from stqdm import stqdm
+
+import src.config as cf
+
+download('punkt', quiet=True)
+wrapper = TextWrapper(cf.MAX_CHAR_LEN, fix_sentence_endings=True)
+
+
+def preprocess(file):
```
Comment (rkingery) [inline src/file_readers.py]: What exactly are you preprocessing? A better name would be <verb describing what you're doing>_<what you're applying the verb to>. In this case, `preprocess_data` or `preprocess_text` or something like that.

File: src/output.py [hunk 16]
```diff
@@ -0,0 +1,79 @@
+"""
+Notes
+-----
+This module contains the functions for audiobook_gen that take the generated audio tensors and output to audio files,
+as well as assembling the final zip archive for user download.
+"""
+
+__all__ = ['write_audio', 'assemble_zip']
```
Comment (rkingery) [inline src/output.py]: You shouldn't need this. See other file for more comments.

File: src/predict.py [hunk 17]
```diff
@@ -0,0 +1,114 @@
+"""
+Notes
+-----
+This module contains the functions for audiobook_gen that handle text-to-speech generation.
+The functions take in the preprocessed text and invoke the Silero package to generate audio tensors.
+"""
+
+__all__ = ['load_model', 'generate_audio', 'predict']
```
Comment (rkingery) [inline src/predict.py]: You shouldn't need this. See other file for more comments.

File: src/output.py [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+    -------
+    zip_name : str
+        name and path of zip directory generated
+
+    """
+    from pathlib import Path
+    import zipfile
+    from stqdm import stqdm
+
+    if not os.path.exists('outputs/'):
```
Comment (rkingery) [inline src/output.py]: You can and should use the pathlib library to perform this action, as it's more modern, safe, and more readable. The `Path` object has both an `exists` method as well as a `mkdir` method. Also, remember not to hardcode file names. Put them in the config.

File: src/output.py [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+        os.mkdir('outputs/')
+
+    directory = Path("outputs/")
+    zip_name = Path(f"outputs/{title}.zip")
```
Comment (rkingery) [inline src/output.py]: You're hardcoding files. Don't do this. Put them in the config and load them.

File: tests/test_file_readers.py [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+    corpus = np.array(rd.preprocess(file), dtype=object)
+    file.close()
+
+    assert np.shape(corpus) == (26,)
+    assert corpus[0] == 'Testing Text File'
+    assert corpus[2] == 'Link to generator repo!'
+    assert corpus[20] == 'Here are some Chinese characters: .'
+    assert corpus[22] == 'The vowels: are , , , , , , .'
+    assert corpus[24] == 'We can also test for mathematical symbols: , , , , , X, %,  ,a, , , +, = ,-.'
+    assert corpus[25] == 'Finally, here are some emoticons: .'
```
Comment (rkingery) [inline tests/test_file_readers.py]: Do you need all of these asserts? The text is coming from the same file. Why not just test equality on the entire file? This is neither memory nor time intensive for a file that small.

File: tests/test_output.py [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+    """
+    import torch
+
+    test_path = "tests/data/test_audio.wav"
+    audio_path = "tests/data/test_audio.pt"
+    audio_list = torch.load(audio_path)
+
+    op.write_audio(audio_list, test_path)
+
+    assert Path(test_path).is_file() is True
```
Comment (rkingery) [inline tests/test_output.py]: The `is True` is redundant. Just type `assert Path(test_path).is_file()`. This format already assumes the variable is a Boolean.

File: tests/test_output.py [hunk 22]
```diff
@@ -0,0 +1,49 @@
...
+def test_assemble_zip():
+    """
+    Tests assemble_zip function, which collects all the audio files from the output directory,
+    and zips them up into a zip directory.
+    """
+    import os
+    from shutil import copy2
+
+    if not os.path.exists('outputs/'):
+        os.mkdir('outputs/')
```
Comment (rkingery) [inline tests/test_output.py]: Use the pathlib library to do this. See other file with similar comment.

File: tests/test_output.py [hunk 23]
```diff
@@ -0,0 +1,49 @@
...
+
+    title = "speaker_samples"
+    direct_res = Path("resources/")
+    direct_out = Path("outputs/")
```
Comment (rkingery) [inline tests/test_output.py]: Hardcoding paths. Don't do this. Use config, or create a separate test_config inside of tests.

File: tests/test_output.py [hunk 24]
```diff
@@ -0,0 +1,49 @@
...
+    for file_path in direct_res.iterdir():
+        if Path(file_path).suffix == '.wav':
+            copy2(file_path, direct_out)
+
+    _ = op.assemble_zip(title)
+
+    assert Path("outputs/speaker_samples.zip").is_file() is True
+    assert Path("outputs/speaker_en_0.wav").is_file() is False
+    assert Path("outputs/speaker_en_110.wav").is_file() is False
```
Comment (rkingery) [inline tests/test_output.py]: These aren't pythonic. Use
`assert Path("outputs/speaker_samples.zip").is_file()`,
`assert not Path("outputs/speaker_en_0.wav").is_file()`
format instead.

Comment (rkingery) [reply tests/test_output.py]: Also, `Path('outputs')` should go in the config instead of being hardcoded. If it's named, say, `outputs_path`, you can create the new filepaths above via
```python
outputs_path / 'speaker_samples.zip'
outputs_path / 'speaker_en_0.wav'
```

File: tests/test_predict.py [hunk 25]
```diff
@@ -0,0 +1,63 @@
...
+import numpy as np
+from pathlib import Path
+
+import src.predict as pr
+import src.file_readers as rd
```
Comment (rkingery) [inline tests/test_predict.py]: This is less readable than just doing `from src import predict` and using `predict.blah` instead of `pr.blah`. You in 2 years will completely forget what `pr` or `rd` mean when reading the code below. Be kind to you in 2 years.

File: tests/test_predict.py [hunk 26]
```diff
@@ -0,0 +1,63 @@
...
+
+
+def test_load_model():
+    """
+    Tests load_model function, which loads the silero TTS model.
+    """
+    model = pr.load_model()
+
+    assert str(type(model)) == "<class '<torch_package_0>.multi_acc_v3_package.TTSModelMultiAcc_v3'>"
```
Comment (rkingery) [inline tests/test_predict.py]: This is a bad way to test type equality. Use the `isinstance` function. Here, e.g., that would look like:
`assert isinstance(model, <whatever the hell the class is named>)`

File: tests/test_predict.py [hunk 27]
```diff
@@ -0,0 +1,63 @@
...
+    assert model.speakers[0] == 'en_0'
+    assert np.shape(model.speakers) == (119,)
+
+
+def test_generate_audio():
+    """
+    Tests generate_audio function, which takes the TTS model and file input,
+    and uses the predict & write_audio functions to output the audio file.
+    """
+    ebook_path = "tests/data/test.epub"
```
Comment (rkingery) [inline tests/test_predict.py]: Hardcoded paths in this file. Don't do this. Use configs for the folder part. See other comments.

File: tests/test_predict.py [hunk 28]
```diff
@@ -0,0 +1,63 @@
...
+    corpus, title = rd.read_epub(ebook_path)
+
+    model = pr.load_model()
+    speaker = 'en_110'
+    pr.generate_audio(corpus[0:2], title, model, speaker)
+
+    assert Path("outputs/the_picture_of_dorian_gray_part000.wav").is_file() is True
+    assert Path("outputs/the_picture_of_dorian_gray_part001.wav").is_file() is True
+    assert Path("outputs/the_picture_of_dorian_gray_part002.wav").is_file() is False
```
Comment (rkingery) [inline tests/test_predict.py:34]: Bad way to write these asserts. See other comments.

File: tests/test_predict.py [hunk 29]
```diff
@@ -0,0 +1,63 @@
...
+
+    Path("outputs/the_picture_of_dorian_gray_part000.wav").unlink()
+    Path("outputs/the_picture_of_dorian_gray_part001.wav").unlink()
```
Comment (rkingery) [inline tests/test_predict.py:37]: Hardcoded paths...

File: tests/test_predict.py [hunk 30]
```diff
@@ -0,0 +1,63 @@
...
+    and appends them together along with a generated file path for output.
+    """
+    import torch
+
+    seed = 1337
+    torch.manual_seed(seed)
+    torch.cuda.manual_seed(seed)
+    model = pr.load_model()
+
+    tensor_path = "tests/data/test_predict.pt"
```
Comment (rkingery) [inline tests/test_predict.py:52]: Hardcoded path...

File: tests/test_predict.py (hunk 28 above)
Comment (rkingery) [inline tests/test_predict.py]: Bad way to write these asserts. See other comments.

**Review by rkingery — CHANGES_REQUESTED** [review]
See comments.

File: src/file_readers.py (hunk 14 above)
Comment (mkutarna) [reply src/file_readers.py]: Understood - I've removed the `__all__` from the source files.

File: src/output.py (hunk 16 above)
Comment (mkutarna) [reply src/output.py]: Removed `__all__`.

File: src/file_readers.py (hunk 15 above)
Comment (mkutarna) [reply src/file_readers.py]: Renamed `preprocess` to `preprocess_text`.

File: src/predict.py (hunk 17 above)
Comment (mkutarna) [reply src/predict.py]: Removed `__all__`.

File: tests/test_output.py (hunk 21 above)
Comment (mkutarna) [reply tests/test_output.py]: Understood - this has been changed here and in the other lines that were written the same way.

File: tests/test_predict.py (hunk 28 above)
Comment (mkutarna) [reply tests/test_predict.py]: Changed as per comments.

File: resources/instructions.md (hunk 13 above)
Comment (mkutarna) [reply resources/instructions.md:1]: This markdown file it used for the streamlit app - I will keep it separate from the README.

As suggested, a link to the HuggingFace Space is in the README, as well as simple instructions for the app.

File: tests/test_predict.py (hunk 25 above)
Comment (mkutarna) [reply tests/test_predict.py]: Implemented this change in all `.py` files.

File: tests/test_predict.py (hunk 27 above)
Comment (mkutarna) [reply tests/test_predict.py]: Fixed this pathing using pathlib instead of hardcoding.

File: tests/test_output.py (hunk 24 above)
Comment (mkutarna) [reply tests/test_output.py]: Fixed this pathing using pathlib instead of hardcoding. Also changed how the asserts are written to following the above format.

File: tests/test_output.py (hunk 23 above)
Comment (mkutarna) [reply tests/test_output.py]: Fixed this pathing using pathlib instead of hardcoding (created `test_config.py` in tests folder).

File: tests/test_output.py (hunk 22 above)
Comment (mkutarna) [reply tests/test_output.py]: Implemented change from `os` to `pathlib`.

File: src/output.py (hunk 19 above)
Comment (mkutarna) [reply src/output.py]: Fixed this pathing using pathlib instead of hardcoding.

File: src/output.py (hunk 18 above)
Comment (mkutarna) [reply src/output.py]: Implemented change to `pathlib` instead of `os`, and using paths instead of hardcoding file names.

File: tests/test_file_readers.py (hunk 20 above)
Comment (mkutarna) [reply tests/test_file_readers.py]: Created a txt file to check equality, rather than all these asserts.

File: tests/test_predict.py (hunk 26 above)
Comment (mkutarna) [reply tests/test_predict.py]: This appears to be a custom message, that doesn't play nicely with the `isinstance` function. For now, I am testing different aspects of the model (e.g. that is contains a select speaker), but I will look into better ways to test the class of the model in the future. Removed that line for the time being.
